core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -4121,6 +4121,102 @@ private void rewriteCountDistinct(QueryModel model) throws SqlException {
         }
     }
 
+    /**
+     * Rewrites "select distinct a,b,c from tab" into "select a,b,c from (select a,b,c,d,count() from tab)".
+     * We will not rewrite distinct of group-by or windows functions.
+     *
+     * @param model input model, it will be validated for the rewrite possibility. If rewrite is impossible,
+     *              the input model will be returned verbatim.
+     * @return either the input model or newly rewritten model.
+     */
+    private QueryModel rewriteDistinct(QueryModel model) throws SqlException {
+        if (model == null) {
+            return null;
+        }
+
+        if (model.isDistinct()) {
+            // bingo
+            // create wrapper models
+            QueryModel wrapperNested = queryModelPool.next();
+            wrapperNested.setNestedModel(model);
+            QueryModel wrapperModel = queryModelPool.next();
+            wrapperModel.setNestedModel(wrapperNested);
+
+            ObjList<QueryColumn> bottomUpColumns = model.getBottomUpColumns();
+            // Columns might be aliased, if they are, we have to create a new
+            // column instance. We also check if we should abandon the rewrite
+            // in case we find counterproductive
+            boolean abandonRewrite = false;
+            for (int i = 0, n = bottomUpColumns.size(); i < n; i++) {
+                QueryColumn qc = bottomUpColumns.getQuick(i);
+                ExpressionNode ast = qc.getAst();
+                CharSequence alias = qc.getAlias();
+                if (qc.isWindowColumn() || (ast.type == FUNCTION && functionParser.getFunctionFactoryCache().isGroupBy(ast.token))) {
+                    abandonRewrite = true;
+                    break;
+                }
+                if (alias == ast.token && ast.type != FUNCTION) {
+                    wrapperModel.addBottomUpColumn(qc);
+                } else {
+                    wrapperModel.addBottomUpColumn(queryColumnPool.next().of(alias, nextLiteral(alias)));
+                }
+            }
+
+            if (!abandonRewrite) {
+
+                // remove the distinct flag, model is not longer that
+                model.setDistinct(false);
+
+                // if we add "count" aggregate we will transform our model pair into
```

Comment:
Moving feedback from #5473 here

> rewrites SQL such as:
> 
> ```sql
> select distinct a, b, c from tab
> ```
> 
> to effectively:
> 
> ```sql
> select a, b, c from (select a, b, c, count() from tab)
> ```

Why not simply
```sql
select a, b, c from tab group by a, b, c;
```
?

Each aggregate function used in the query adds some overhead and we don't use that `count()` in the end.

Reply:
you have to try queries before and after. SQL such as 
```sql
select distinct symbol from tab
```

will use "rosti", if you don't add count - it will stay in Java land 2x slower. I added "null_l()" function, which is a noop, however,

```sql
select symbol, count() ...
```

is still faster than
```
select symbol, null_l()
```

I'll mess around with `null_l` more to see where it is faster, but that would not be uniform.

Ultimately this rewrite with unused "count" is still at least 10-100x faster than the existing "distinct" implementation

Reply:
yeah, there is basically there no difference between:

```sql
select  symbol, side, null_l() from trades;
and
select  symbol, side, count() from trades;
```

keying the map dominates the execution

Reply:
On demo
```sql
EXPLAIN SELECT symbol FROM trades GROUP BY symbol;
```
gives

QUERY PLAN
--
GroupBy vectorized: true workers: 46
&nbsp;&nbsp;keys: [symbol]
&nbsp;&nbsp;values: [count(*)]
&nbsp;&nbsp;&nbsp;&nbsp;PageFrame
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: trades

so rosti factories inject count() on their own for key-only queries (which is something we could fix by proper rosti hash set in future BTW). Am I missing something?

> it will stay in Java land 2x slower

A side note: if there are many distinct key values, rosti will be 5-10x slower as it lacks high cardinality handling.

Reply:
i think this is a separate problem, we can address this separately.

I think "count()" is irrelevant, you can compare perf of these two SQLs on demo:
```sql
select symbol, side from trades group by symbol, side;
select symbol, side, count() from trades group by symbol, side;
```

outside of this PR I would address "count" concern as follows:
- impl rosti-based no-op aggregate and see if this moves a needle
- reduce writes by non-rosti maps and see if removing count() is this case moves a needle
- remove count() when it make a difference

Reply:
> I think "count()" is irrelevant, you can compare perf of these two SQLs on demo

Here is what I got on demo (best hot exec time):
```sql
-- 1.13s
select symbol, side from trades group by symbol, side;
-- 1.68s
select symbol, side, count() from trades group by symbol, side;
```

So, it seems to make a difference.

Reply:
it really isn't easy to get 1.13s on demo for query 1, my mileage is different, more like 1.6 and about that for the query 2, but 1.8s mostly.

I looked into removing count() - it will be a pain to do, right now adding "count" avoids the rewrite to deal with * expansion. If we add columns to the group-by list - rewrite will have to deal with * expansion, including aliased *, joins, and the pre-existing group-by columns - the full hog.

Just to put this PR in the context, we're going from 38s on 
```
select distinct symbol, side from trades
```

to 1.8s after.

We can squeeze a little more out of it separately i think.

Reply:
OK, if improving the rewrite and getting rid of this redundant function is tricky, we could leave it as is and, maybe, create a GH issue.